File: atividades/exercicios_fixacao/Aula_5/C++/polimorfismo_aula_5.cpp
#include <iostream>
#include <string>
#include <vector>
#include <memory>

class Veiculo {
protected:
    std::string modelo;

public:
    Veiculo (std::string m) 
        : modelo(m) {}

    virtual void acelerar() const = 0;

    virtual ~Veiculo() = default;
};

class Carro : public Veiculo {
public:
    Carro (std::string m) 
        : Veiculo(m) {}

    void acelerar() const override {
        std::cout << "O " << modelo << " acelera pelo pedal \n";
    }
};

class CarroEletrico : public Veiculo {
public:
    CarroEletrico (std::string m) 
        : Veiculo(m) {}

    void acelerar() const override {
        std::cout << "O " << modelo << " acelera rapido e nao faz barulho \n";
    }
};

class Moto : public Veiculo {
public:
    Moto (std::string m) 
        : Veiculo(m) {}

    void acelerar() const override {
        std::cout << "A " << modelo << " acelera muito rapido e tem duas rodas \n";
    }
};

class Caminhao : public Veiculo {
public:
    Caminhao (std::string m)
        : Veiculo(m) {}

    void acelerar() const override {
        std::cout <<  "O " << modelo << " e lento mas muito forte \n";
    }
};

int main() {

    std::vector<std::unique_ptr<Veiculo>> veiculos;

    veiculos.push_back(std::make_unique<Carro>("Carro"));
    veiculos.push_back(std::make_unique<CarroEletrico>("Carro eletrico"));
    veiculos.push_back(std::make_unique<Moto>("Moto"));
    veiculos.push_back(std::make_unique<Caminhao>("Caminhao"));

    for (const auto& v : veiculos) {
        v->acelerar();
    }


    return 0;
}

File: atividades/exercicios_fixacao/Aula_5/polimorfismo_aula_5.py
from abc import ABC, abstractmethod

class Veiculo:
    def __init__(self, modelo):
        self.modelo = modelo

    @abstractmethod
    def acelerar(self):
        pass

class Carro(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)
        
    def acelerar(self):
        print(f"{self.modelo} acelera pelo pedal")

class CarroEletrico(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

    def acelerar(self):
        print(f"O carro eletrco acelera pelo pedal mas não faz barulho")

class Moto(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

    def acelerar(self):
        print(f"{self.modelo}, acelera pela manete")

class Caminhao(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

    def acelerar(self):
        print(f"{self.modelo} acelerou devagar mas é muito forte")

veiculos = [
    Carro("Carro"),
    CarroEletrico("Carro elétrico"),
    Moto("Moto"),
    Caminhao("Caminhão")
]

for i in veiculos:
    i.acelerar()